MonoGame.Framework/Quaternion.cs

```diff
@@ -61,242 +147,331 @@ public static Quaternion Add(Quaternion quaternion1, Quaternion quaternion2)
-		        float num3 = 0.5f / num6;
-		        result.X = (matrix.M21 + matrix.M12) * num3;
-		        result.Y = 0.5f * num6;
-		        result.Z = (matrix.M32 + matrix.M23) * num3;
-		        result.W = (matrix.M31 - matrix.M13) * num3;
-		    }
-		    else
-		    {
-		        float num5 = (float) Math.Sqrt((double) (((1f + matrix.M33) - matrix.M11) - matrix.M22));
-		        float num2 = 0.5f / num5;
-		        result.X = (matrix.M31 + matrix.M13) * num2;
-		        result.Y = (matrix.M32 + matrix.M23) * num2;
-		        result.Z = 0.5f * num5;
-		        result.W = (matrix.M12 - matrix.M21) * num2;
-		    }
+            float sqrt;
+            float half;
+            float scale = matrix.M11 + matrix.M22 + matrix.M33;
+
+            if (scale > 0.0f)
+            {
+                sqrt = (float)Math.Sqrt(scale + 1.0f);
+                result.W = sqrt * 0.5f;
+                sqrt = 0.5f / sqrt;
+
+                result.X = (matrix.M23 - matrix.M32) * sqrt;
+                result.Y = (matrix.M31 - matrix.M13) * sqrt;
+                result.Z = (matrix.M12 - matrix.M21) * sqrt;
+            }
+            else
+            if ((matrix.M11 >= matrix.M22) && (matrix.M11 >= matrix.M33))
+            {
+                sqrt = (float)Math.Sqrt(1.0f + matrix.M11 - matrix.M22 - matrix.M33);
+                half = 0.5f / sqrt;
 
+                result.X = 0.5f * sqrt;
+                result.Y = (matrix.M12 + matrix.M21) * half;
+                result.Z = (matrix.M13 + matrix.M31) * half;
+                result.W = (matrix.M23 - matrix.M32) * half;
+            }
+            else if (matrix.M22 > matrix.M33)
+            {
+                sqrt = (float) Math.Sqrt(1.0f + matrix.M22 - matrix.M11 - matrix.M33);
+                half = 0.5f/sqrt;
+
+                result.X = (matrix.M21 + matrix.M12)*half;
+                result.Y = 0.5f*sqrt;
+                result.Z = (matrix.M32 + matrix.M23)*half;
+                result.W = (matrix.M31 - matrix.M13)*half;
+            }
+            else
+            {
+                sqrt = (float)Math.Sqrt(1.0f + matrix.M33 - matrix.M11 - matrix.M22);
+                half = 0.5f / sqrt;
+
+                result.X = (matrix.M31 + matrix.M13) * half;
+                result.Y = (matrix.M32 + matrix.M23) * half;
+                result.Z = 0.5f * sqrt;
+                result.W = (matrix.M12 - matrix.M21) * half;
+            }
         }
-		
-		public static Quaternion CreateFromYawPitchRoll(float yaw, float pitch, float roll)
+
+        #endregion
+
+        #region CreateFromYawPitchRoll
+
+        /// <summary>
+        /// Creates a new <see cref="Quaternion"/> from the specified yaw, pitch and roll scalars.
+        /// </summary>
+        /// <param name="yaw"></param>
+        /// <param name="pitch"></param>
+        /// <param name="roll"></param>
+        /// <returns>The new quaternion builded from the specified yaw, pitch and roll scalars.</returns>
```

Comment:
Creates a new <see cref="Quaternion"/> from the specified yaw, pitch and roll scalars. - should I change it too ?